### DISCUSSION TITLE
How should the SMO nuget packages be structured?

### DISCUSSION THREAD

**Discussion by shueybubbles** [microsoft/sqlmanagementobjects#46] [Ideas]
Historically, the "SQL Management Objects" label has been an umbrella term for a collection of DLLs that is always installed as an intact set. For years we distributed it via an MSI.

Now that SMO is available through [nuget](https://www.nuget.org/packages/Microsoft.SqlServer.SqlManagementObjects/)  and the nuget package managers for VS2019+ are powerful, we have more options available for structuring these packages. 

I am looking for community feedback on the current and future structure of the SMO nuget package to help prioritize any investment in this area over the coming year.

## Current structure 

Today, we have a single SMO package that doubles as both the compilation reference and the binary distribution package to be consumed by applications such as the [SQL Tools Service](https://github.com/microsoft/sqltoolsservice). We optimized the structure for time-to-market when we moved the SMO build to a VS2019 retail msbuild system and we didn't want to spend time creating any fine grained packages. 

The first question I have is whether SMO application developers are even concerned about this or if the current monolithic package is satisfying everyone's needs already. If you are happy with the status quo, please let us know that! If you feel like the current package structure hinders your ability to create applications, please consider the alternatives outline below and provide feedback.

thx

### Option 1- Keep it "as is",

Pros: 

- No need to spend time changing the build or any of the SMO consumers.
- Lessens the chances of dependency inclusion problems for consumers as we see with [Microsoft.Data.SqlClient.SNI](https://github.com/dotnet/SqlClient/issues/446) today. If you are using SMO with older build systems using packages.config file, having everything you need in one package likely simplifies your life.
- As new functionality becomes available, it's all discoverable in this single package. 
 
Cons: 

- The package is likely bigger than most consumers need or prefer.
- Consumers that want to try to minimize the size of their own app need to identify on their own which components of the package are safe to omit.
- During SMO development it's too tempting/easy to add dependencies between DLLs that may not be appropriate.

### Option 2 - Put each DLL in its own package and change Microsoft.SqlServer.SqlManagementObjects to simply pull in all the DLL packages as a dependency.

We can rely on "dotnet pack" to create a nupkg for each DLL which will automatically include compile-time dependencies as dependencies in the nuspec for each generated package.

Pros:

- Consumers can use just the components they need to minimize app size. Some classes in ConnectionInfo in particular can be useful on their own, such as ServerConnection.

Cons: 

- Apps using packages.config will have to manage all the dependency identification on their own.
- Not every dependency in SMO is compile-time. Some are dynamic/reflection-based. As SMO evolves these dependencies will have to be accounted for to make package generation have the correct dependencies.

### Option 3 - Break the package up into a handful of smaller packages, partitioned by functionality.

There are several possible groupings of SMO DLLs.  For example, we could have a Microsoft.SqlServer.SqlManagementObjects.Core package which only includes these DLLs:

```
Microsoft.SqlServer.SqlEnum.dll
Microsoft.SqlServer.RegSvrEnum.dll
Microsoft.SqlServer.Management.Sdk.Sfc.dll
Microsoft.SqlServer.Dmf.dll
Microsoft.SqlServer.Dmf.Common.dll
Microsoft.Data.Tools.Sql.BatchParser.dll
Microsoft.SqlServer.ConnectionInfo.dll
Microsoft.SqlServer.Smo.dll
Microsoft.SqlServer.ServiceBrokerEnum.dll
```

These DLLs provide the functionality for scripting most SQL objects and the vast majority of the overall object model. We could put the DLLs for XEvents/WMI/Collector in a "Miscellaneous" package, SqlScriptPublish and Notebook in one, SmoExtended and SmoMetadataProvider in their own, etc.

Pros:

- Package consumers can include support for just the scenarios they need in their application.
- We minimize the dependency juggling required by apps built using packages.config if they don't need all of SMO.

Cons: 

- Our choice of DLL groupings might not reflect real world consumption patterns.
- Creating and maintaining the build for the packages is more complicated than either option 1 or option 2.

**Comment by Matteo-T** [comment]
Though choice!

Personally, I like `Option 1` - because of all the pros you listed. And I have hope that we may be able to use the "new" Trim functionality in .Net... (perhaps when we replace all the reflection with source generators?)... so we delegate the "please make my app smaller" to the compiler.

Alternatively, we could go `Option 3` - a compromise could be to have `small` number of packages (maybe 2... like a new ".core" package and an "ext" one (for stuff like WMI, XEvent, ....; ."ext" would be what we have today with a dependency on "."core").

We need more people to comment here! :)

**Comment by potatoqualitee** [comment]
Understanding libraries is one of my weak points. I know how to use it but I don't know how it works.

I generally need packaging as simple as possible and I may be an exception in this area, I'd go for "as is" because I can scrape by with my experience. Any others, and I'd need to consult with someone to tell me how to combine everything into something that works for dbatools. And it especially hurts because `Install-Package` does not work unless `-IgnoreDependencies` is specified so I'd have to go find everything myself.

Option 2 seems the worst for me because it's hard for me to resolve anything that's dynamic. I had this problem with Go and had to do something static and eventually gave up. I like how it is now even with the size because I know it works. But I'm open to others so long as there's some way to make it work like it does now. I also do love smaller.

Perhaps @FriedrichWeinmann has some insight as to which one would be most helpful for us. He's our C# guy.

**Comment by FriedrichWeinmann** [comment]
Well, I'd go for Option 4:
Do both 1 and 2:

- Release the full package for ease of use, as previously
- Also release individual packages, for those needing to minimize their image size (think container deployments or function apps).

This solves both scenarios, does not break anything for anybody _and_ forces you to be clear with your dependency management and keeping your code sane.
With invisible, dynamic dependencies and delivering it as one big package, you are making life ... interesting for anybody trying to pare down the current package to what they actually need.
